File: eval/run_eval.py
#!/usr/bin/env python3
"""Benchmark a local/free Indonesian<->English translation backend.

Examples:
  # Ollama (TranslateGemma uses its own recommended prompt template)
  python3 eval/run_eval.py --backend ollama --model translategemma:4b \
      --prompt-style translate_gemma --src id --tgt en --name translategemma-4b

  # Any OpenAI-compatible server (LM Studio, llama-server, vLLM)
  python3 eval/run_eval.py --backend openai --model Hy-MT2-1.8B \
      --base-url http://127.0.0.1:8080/v1 --prompt-style hymt2 --src id --tgt en

  # Dedicated NMT via transformers (NLLB / M2M-100 / MADLAD / OPUS-MT)
  python3 eval/run_eval.py --backend transformers --model facebook/nllb-200-distilled-600M \
      --family nllb --prompt-style none --src id --tgt en

  # Cloudflare Workers AI (official SDK; needs CLOUDFLARE_API_TOKEN +
  # CLOUDFLARE_ACCOUNT_ID in the environment, i.e. `set -a; . ./.env; set +a`)
  python3 eval/run_eval.py --backend cloudflare --model @cf/meta/m2m100-1.2b \
      --prompt-style none --src id --tgt en --name cf-m2m100-1.2b

  # Kagi Translate via bevry-vibes/kagi-translate-client (needs KAGI_SESSION and
  # KAGI_CLIENT_REPO in the environment; --kagi-runtime python|deno picks the CLI)
  python3 eval/run_eval.py --backend kagi --kagi-runtime python \
      --prompt-style none --src id --tgt en --name kagi-py

Hosted backends (cloudflare, kagi) load no local model, so the memguard RAM fit
check is skipped for them; the one-benchmark-at-a-time run lock still applies.
Raw results land in results/*.json; `python3 eval/summarize.py` renders results/results.md.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

import memguard  # noqa: E402
import metrics  # noqa: E402
from backends import (  # noqa: E402
    ArgosBackend, CloudflareBackend, KagiBackend, OllamaBackend, OpenAICompatBackend,
    TransformersBackend,
)

ROOT = Path(__file__).resolve().parent.parent
DEFAULT_TESTSET = ROOT / "data" / "testset.jsonl"
RESULTS = ROOT / "results"


def load_pairs(path: Path, src: str, tgt: str) -> list[dict]:
    pairs = []
    for line in path.read_text(encoding="utf-8").splitlines():
        if not line.strip():
            continue
        entry = json.loads(line)
        if entry["src"] == src and entry["tgt"] == tgt:
            pairs.append(entry)
    return pairs


def make_backend(args):
    if args.backend == "ollama":
        think = None if args.think is None else args.think
        return OllamaBackend(args.model, host=args.host, prompt_style=args.prompt_style,
                             temperature=args.temperature, num_ctx=args.num_ctx, think=think)
    if args.backend == "openai":
        return OpenAICompatBackend(args.model, base_url=args.base_url, api_key=args.api_key,
                                   prompt_style=args.prompt_style, temperature=args.temperature)
    if args.backend == "cloudflare":
        token = args.api_key if args.api_key != "not-needed" else os.environ.get("CLOUDFLARE_API_TOKEN", "")
        account = args.account_id or os.environ.get("CLOUDFLARE_ACCOUNT_ID", "")
        if not token or not account:
            raise SystemExit("cloudflare backend needs CLOUDFLARE_API_TOKEN and CLOUDFLARE_ACCOUNT_ID "
                             "in the environment (`set -a; . ./.env; set +a`) or --api-key/--account-id")
        return CloudflareBackend(args.model, account_id=account, api_token=token)
    if args.backend == "kagi":
        if not os.environ.get("KAGI_SESSION"):
            raise SystemExit("kagi backend needs KAGI_SESSION in the environment "
                             "(`set -a; . ./.env; set +a`)")
        return KagiBackend(runtime=args.kagi_runtime, kagi_client_repo=args.kagi_client_repo)
    if args.backend == "transformers":
        return TransformersBackend(args.model, family=args.family)
    if args.backend == "argos":
        return ArgosBackend()
    raise SystemExit(f"unknown backend: {args.backend}")


def by_category(pairs: list[dict], per_sentence: list[float]) -> dict:
    buckets: dict[str, list[float]] = {}
    for entry, score in zip(pairs, per_sentence):
        buckets.setdefault(entry["category"], []).append(score)
    return {cat: round(sum(vals) / len(vals), 2) for cat, vals in sorted(buckets.items())}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--backend", required=True,
                        choices=["ollama", "openai", "transformers", "argos", "cloudflare", "kagi"])
    parser.add_argument("--model", default="")
    parser.add_argument("--family", default="seq2seq",
                        help="transformers family: nllb | m2m100 | madlad | opus | seq2seq")
    parser.add_argument("--prompt-style", default="generic",
                        choices=["generic", "translate_gemma", "hymt2", "none"])
    parser.add_argument("--src", default="id")
    parser.add_argument("--tgt", default="en")
    parser.add_argument("--testset", default=str(DEFAULT_TESTSET))
    parser.add_argument("--limit", type=int, default=0, help="cap number of pairs (0 = all)")
    parser.add_argument("--name", default=None, help="label used in result filenames")
    parser.add_argument("--host", default="http://127.0.0.1:11434")
    parser.add_argument("--base-url", default="http://127.0.0.1:1234/v1")
    parser.add_argument("--api-key", default="not-needed")
    parser.add_argument("--account-id", default=None,
                        help="Cloudflare account id (default $CLOUDFLARE_ACCOUNT_ID)")
    parser.add_argument("--kagi-runtime", default="python", choices=["python", "deno"],
                        help="which kagi-translate-client CLI to drive (default python)")
    parser.add_argument("--kagi-client-repo", default=None,
                        help="path to bevry-vibes/kagi-translate-client (default $KAGI_CLIENT_REPO)")
    parser.add_argument("--temperature", type=float, default=0.0)
    parser.add_argument("--num-ctx", type=int, default=2048)
    parser.add_argument("--think", dest="think", action="store_true", default=None,
                        help="let thinking models (Qwen3, ...) emit reasoning before the translation")
    parser.add_argument("--no-think", dest="think", action="store_false",
                        help="disable reasoning tokens (recommended for translation)")
    parser.add_argument("--est-gb", type=float, default=None,
                        help="override the estimated RAM requirement in GB")
    parser.add_argument("--min-free-gb", type=float, default=2.0,
                        help="RAM that must remain free after loading the model (default 2.0)")
    parser.add_argument("--force", action="store_true",
                        help="run even if the memory guard says there is not enough RAM")
    parser.add_argument("--keep-loaded", action="store_true",
                        help="leave the Ollama model in RAM after the run (default: unload it)")
    parser.add_argument("--no-warmup", action="store_true",
                        help="charge model load time to the first sentence instead of pre-warming")
    return parser.parse_args()


def run(args: argparse.Namespace) -> None:
    pairs = load_pairs(Path(args.testset), args.src, args.tgt)
    if args.limit:
        pairs = pairs[: args.limit]
    if not pairs:
        raise SystemExit(f"no {args.src}->{args.tgt} pairs in {args.testset}; "
                         f"run `python3 eval/build_testset.py` first")

    backend = make_backend(args)
    label = args.name or f"{args.backend}-{args.model or 'default'}".replace("/", "_")
    hf = f"{label}-{args.src}{args.tgt}"
    print(f"[{datetime.now(timezone.utc).isoformat(timespec='seconds')}] "
          f"{backend.name} | {len(pairs)} pairs {args.src}->{args.tgt} | prompt={args.prompt_style}")

    # --- memory guards: refuse to start when the box cannot hold this model ----------------
    resident = memguard.resident_ollama_models(args.host) if args.backend == "ollama" else []
    if resident and args.backend == "ollama":
        print(f"[mem ] Ollama already holds in RAM: {', '.join(resident)}")
    if getattr(backend, "loads_local_model", True):
        required = memguard.estimate_gb(args.model or "unknown", args.family,
                                        num_ctx=args.num_ctx, explicit=args.est_gb)
        memguard.require_memory(required, min_free_gb=args.min_free_gb, force=args.force,
                                label=hf)
    else:
        print("[mem ] hosted backend: no local model to load, skipping the RAM fit check "
              "(the run lock is still held)")

    # load the model / warm the KV cache before the clock starts (skippable with --no-warmup)
    if hasattr(backend, "warmup") and not args.no_warmup:
        backend.warmup(args.src, args.tgt)

    sources = [p["source"] for p in pairs]
    references = [p["reference"] for p in pairs]
    started = time.perf_counter()
    result = backend.translate(sources, args.src, args.tgt)
    wall = time.perf_counter() - started

    per_sentence = metrics.per_sentence_chrf(result.texts, references)
    scores = metrics.score_all(result.texts, references)
    payload = {
        "label": hf,
        "backend": args.backend,
        "model": args.model or result.model,
        "prompt_style": args.prompt_style,
        "src": args.src,
        "tgt": args.tgt,
        "pairs": len(pairs),
        "metrics": scores,
        "metric_backend": metrics.metric_backend(),
        "exact_match_rate": metrics.exact_match_rate(result.texts, references),
        "seconds_total": round(wall, 2),
        "seconds_per_sentence": round(wall / len(pairs), 3),
        "chrF_by_category": by_category(pairs, per_sentence),
        "timestamp": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "samples": [
            {"id": p["id"], "category": p["category"], "source": p["source"],
             "reference": p["reference"], "hypothesis": h, "chrf": s}
            for p, h, s in zip(pairs, result.texts, per_sentence)
        ],
    }

    RESULTS.mkdir(parents=True, exist_ok=True)
    out = RESULTS / f"{hf}.json"
    out.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")

    print(f"  bleu={scores['bleu']}  chrF={scores['chrf']}  chrF++={scores['chrfpp']}  "
          f"(metric backend: {payload['metric_backend']})")
    print(f"  wall={payload['seconds_total']}s  {payload['seconds_per_sentence']}s/sentence")
    print("  worst 5 by chrF:")
    for entry, hyp, score in sorted(zip(pairs, result.texts, per_sentence), key=lambda r: r[2])[:5]:
        print(f"   - [{score:6.2f}] {entry['category']}: {entry['source'][:70]!r}")
        print(f"            ref: {entry['reference'][:70]!r}")
        print(f"            hyp: {hyp[:70]!r}")
    print(f"wrote {out}")

    # evict the model so the next run starts from a clean memory state
    if args.backend == "ollama" and not args.keep_loaded and args.model:
        memguard.unload_ollama([args.model], host=args.host)
    print(f"[mem ] {memguard.available_gb():.1f} GB RAM available after unload")
    print("run `python3 eval/summarize.py` to regenerate results/results.md")


def main() -> None:
    # One benchmark at a time, ever: a second concurrent run fails fast here
    # instead of stacking another multi-GB model on top of the resident one.
    with memguard.RunLock():
        run(parse_args())


if __name__ == "__main__":
    main()

File: eval/deno/run_eval.ts
#!/usr/bin/env -S deno run --allow-net --allow-env --allow-read --allow-write
/** Like-for-like Deno harness for the Cloudflare Workers AI translation models.
 *
 * Same testset, same local-metric formulas (./metrics.ts), same results JSON
 * schema and one-benchmark-at-a-time run lock as eval/run_eval.py, so
 * eval/summarize.py renders both clients into one results/results.md.
 *
 * Usage:
 *   deno run --allow-net --allow-env --allow-read --allow-write eval/deno/run_eval.ts \
 *     --model @cf/meta/m2m100-1.2b --src id --tgt en --name cf-m2m100-1.2b-deno
 *
 * Needs CLOUDFLARE_API_TOKEN and CLOUDFLARE_ACCOUNT_ID in the environment
 * (`set -a; . ./.env; set +a`). The typed ai.run() helper URL-encodes the
 * slash-bearing model id (see eval/backends.py CloudflareBackend), so this
 * drives the SDK's generic request path too.
 */

import Cloudflare from "npm:cloudflare@7.1.0";
import { exactMatchRate, perSentenceChrf, scoreAll } from "./metrics.ts";

// Corpus-level scoring mirrors metrics.py exactly: sacrebleu when the system
// python3 has it (one subprocess per run — scoring only, the client stays Deno),
// otherwise the bundled local formulas. This keeps the Deno rows directly
// comparable with the Python rows in results/results.md.
const SACREBLEU_SNIPPET = [
  "import json, sys, sacrebleu",
  "h, r = json.load(sys.stdin)",
  "print(json.dumps({",
  "    'bleu': round(float(sacrebleu.corpus_bleu(h, [r]).score), 2),",
  "    'chrf': round(float(sacrebleu.corpus_chrf(h, [r]).score), 2),",
  "    'chrfpp': round(float(sacrebleu.corpus_chrf(h, [r], word_order=2).score), 2),",
  "}))",
].join("\n");

async function scoreAllParity(
  hypotheses: string[],
  references: string[],
): Promise<{ scores: { bleu: number; chrf: number; chrfpp: number }; backend: string }> {
  try {
    const command = new Deno.Command("python3", {
      args: ["-c", SACREBLEU_SNIPPET],
      stdin: "piped",
      stdout: "piped",
      stderr: "piped",
    });
    const child = command.spawn();
    const writer = await child.stdin.getWriter();
    await writer.write(new TextEncoder().encode(JSON.stringify([hypotheses, references])));
    writer.close();
    const stdout = await new Response(child.stdout).text();
    const status = await child.status;
    if (status.success) {
      const scores = JSON.parse(stdout);
      return { scores, backend: "sacrebleu" };
    }
  } catch {
    // fall through to the bundled local formulas
  }
  return { scores: scoreAll(hypotheses, references), backend: "local" };
}

const ROOT = new URL("../../", import.meta.url).pathname;
const DEFAULT_TESTSET = ROOT + "data/testset.jsonl";
const RESULTS = ROOT + "results";
const LOCK_PATH = Deno.env.get("MTBENCH_LOCK") ?? "/tmp/indonesian-mt-bench.lock";

interface Pair {
  id: string;
  category: string;
  source: string;
  reference: string;
}

function parseArgs(argv: string[]): Record<string, string> {
  const opts: Record<string, string> = {};
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (!arg.startsWith("--")) continue;
    const key = arg.slice(2);
    if (i + 1 < argv.length && !argv[i + 1].startsWith("--")) {
      opts[key] = argv[++i];
    } else {
      opts[key] = "true";
    }
  }
  return opts;
}

function loadPairs(path: string, src: string, tgt: string): Pair[] {
  const pairs: Pair[] = [];
  for (const line of Deno.readTextFileSync(path).split("\n")) {
    if (!line.trim()) continue;
    const entry = JSON.parse(line);
    if (entry.src === src && entry.tgt === tgt) pairs.push(entry);
  }
  return pairs;
}

// Mirror of memguard.RunLock: fail fast when another live pid holds the lock;
// treat the lock as stale when the pid is dead or the file is older than 3 hours.
function acquireLock(): void {
  try {
    const stat = Deno.statSync(LOCK_PATH);
    const pid = parseInt(Deno.readTextFileSync(LOCK_PATH).trim(), 10);
    let alive = false;
    try {
      Deno.statSync(`/proc/${pid}`);
      alive = true;
    } catch {
      alive = false;
    }
    const ageHours = (Date.now() - (stat.mtime?.getTime() ?? 0)) / 3_600_000;
    if (alive && ageHours < 3) {
      console.error(`another benchmark holds the lock (pid ${pid}): ${LOCK_PATH}`);
      Deno.exit(1);
    }
  } catch (e) {
    if (!(e instanceof Deno.errors.NotFound)) throw e;
  }
  Deno.writeTextFileSync(LOCK_PATH, String(Deno.pid));
}

function releaseLock(): void {
  try {
    Deno.removeSync(LOCK_PATH);
  } catch (e) {
    if (!(e instanceof Deno.errors.NotFound)) throw e;
  }
}

function extractTranslation(result: Record<string, unknown>, model: string): string {
  if (typeof result.translated_text === "string") { // m2m100-style: a single string
    return result.translated_text.trim();
  }
  if (Array.isArray(result.translations)) { // indictrans2-style: a list of segments
    const parts = (result.translations as unknown[]).map((p) =>
      typeof p === "string" ? p : String((p as Record<string, unknown>)?.translated_text ?? "")
    );
    return parts.join(" ").trim();
  }
  throw new Error(`cloudflare: unexpected result shape from ${model}: ${JSON.stringify(result).slice(0, 200)}`);
}

function byCategory(pairs: Pair[], perSentence: number[]): Record<string, number> {
  const buckets = new Map<string, number[]>();
  pairs.forEach((pair, i) => {
    const arr = buckets.get(pair.category) ?? [];
    arr.push(perSentence[i]);
    buckets.set(pair.category, arr);
  });
  const out: Record<string, number> = {};
  for (const cat of [...buckets.keys()].sort()) {
    const vals = buckets.get(cat)!;
    out[cat] = Math.round((vals.reduce((a, b) => a + b, 0) / vals.length) * 100) / 100;
  }
  return out;
}

async function main() {
  const opts = parseArgs(Deno.args);
  const model = opts.model;
  if (!model) throw new Error("this harness needs --model (e.g. @cf/meta/m2m100-1.2b)");
  const src = opts.src ?? "id";
  const tgt = opts.tgt ?? "en";
  const apiToken = opts["api-key"] ?? Deno.env.get("CLOUDFLARE_API_TOKEN") ?? "";
  const accountId = opts["account-id"] ?? Deno.env.get("CLOUDFLARE_ACCOUNT_ID") ?? "";
  if (!apiToken || !accountId) {
    throw new Error("cloudflare needs CLOUDFLARE_API_TOKEN and CLOUDFLARE_ACCOUNT_ID in the environment (`set -a; . ./.env; set +a`)");
  }

  const pairs = loadPairs(opts.testset ?? DEFAULT_TESTSET, src, tgt);
  if (opts.limit) pairs.length = Math.min(pairs.length, parseInt(opts.limit, 10));
  if (pairs.length === 0) {
    throw new Error(`no ${src}->${tgt} pairs in the testset; run \`python3 eval/build_testset.py\` first`);
  }

  const label = opts.name ?? `cloudflare-deno-${model}`;
  const hf = `${label}-${src}${tgt}`;
  console.log(`[${new Date().toISOString().slice(0, 19)}+00:00] cloudflare-deno:${model} | ${pairs.length} pairs ${src}->${tgt} | prompt=none`);

  acquireLock();
  try {
    const client = new Cloudflare({ apiToken });
    const path = `/accounts/${accountId}/ai/run/${model}`;

    // one throwaway call so connection setup is not charged to the benchmark
    await client.post(path, { body: { text: "Halo.", source_lang: src, target_lang: tgt } });

    const sources = pairs.map((p) => p.source);
    const references = pairs.map((p) => p.reference);
    const out: string[] = [];
    const started = performance.now();
    for (const text of sources) {
      const body = await client.post(path, {
        body: { text, source_lang: src, target_lang: tgt },
      }) as Record<string, unknown>;
      if (!body.success) {
        throw new Error(`cloudflare: bad envelope from ${model}: ${JSON.stringify(body).slice(0, 200)}`);
      }
      out.push(extractTranslation((body.result ?? {}) as Record<string, unknown>, model));
    }
    const wallSeconds = (performance.now() - started) / 1000;

    const perSentence = perSentenceChrf(out, references);
    const { scores, backend: metricBackend } = await scoreAllParity(out, references);
    const payload = {
      label: hf,
      backend: "cloudflare-deno",
      model,
      prompt_style: "none",
      src,
      tgt,
      pairs: pairs.length,
      metrics: scores,
      metric_backend: metricBackend,
      exact_match_rate: exactMatchRate(out, references),
      seconds_total: Math.round(wallSeconds * 100) / 100,
      seconds_per_sentence: Math.round(wallSeconds / pairs.length * 1000) / 1000,
      chrF_by_category: byCategory(pairs, perSentence),
      timestamp: new Date().toISOString().slice(0, 19) + "+00:00",
      samples: pairs.map((pair, i) => ({
        id: pair.id,
        category: pair.category,
        source: pair.source,
        reference: pair.reference,
        hypothesis: out[i],
        chrf: perSentence[i],
      })),
    };

    Deno.mkdirSync(RESULTS, { recursive: true });
    const outPath = `${RESULTS}/${hf}.json`;
    Deno.writeTextFileSync(outPath, JSON.stringify(payload, null, 2) + "\n");

    console.log(`  bleu=${scores.bleu}  chrF=${scores.chrf}  chrF++=${scores.chrfpp}  (metric backend: ${metricBackend})`);
    console.log(`  wall=${payload.seconds_total}s  ${payload.seconds_per_sentence}s/sentence`);
    console.log("  worst 5 by chrF:");
    [...pairs.keys()]
      .sort((a, b) => perSentence[a] - perSentence[b])
      .slice(0, 5)
      .forEach((i) => {
        console.log(`   - [${perSentence[i].toFixed(2).padStart(6)}] ${pairs[i].category}: ${JSON.stringify(pairs[i].source.slice(0, 70))}`);
        console.log(`            ref: ${JSON.stringify(references[i].slice(0, 70))}`);
        console.log(`            hyp: ${JSON.stringify(out[i].slice(0, 70))}`);
      });
    console.log(`wrote ${outPath}`);
    console.log("run `python3 eval/summarize.py` to regenerate results/results.md");
  } finally {
    releaseLock();
  }
}

await main();
